docs/using-modules.rst

```diff
@@ -0,0 +1,190 @@
+
+Initializing a module
+=====================
+
+In order to use a module's state, it must be "initialized". A module can be initialized with the ``initializes`` keyword. This declares the module's location in the contract's :ref:`Storage Layout <compiler-storage-layout>`. It also creates a requirement to invoke the module's :ref:`__init__() function <init-function>`, if it has one. This is a well-formedness requirement, since it does not make sense to access a module's state unless its ``__init__()`` function has been called.
+
+.. code-block:: vyper
+
+    import ownable
+
+    initializes: ownable
+
+    @deploy
+    def __init__():
+        ownable.__init__()
+
+    @external
+    def my_access_controlled_function():
+        ownable._check_owner()  # reverts unless msg.sender == ownable.owner
+
+        ... # do things that only the owner can do
+
+It is a compile-time error to invoke a module's ``__init__()`` function more than once!
+
+A module's state can be directly accessed just by prefixing the name of a variable with the module's alias, like follows:
+
+.. code-block:: vyper
+
+    @external
+    def get_owner() -> address:
+        return ownable.owner
+
+
+The ``uses`` statement
+======================
+
+Another way of using a contract's state without directly initializing it is to use the ``uses`` keyword. This is a more advanced usage which is expected to be mostly utilized by library designers. The ``uses`` statement allows a module to use another module's state but defer its initialization to another module in the compilation tree (most likely a user of the library in question).
+
+This is best illustrated with an example:
+
+.. code-block:: vyper
+
+    # ownable_2step.vy
+    import ownable
+
+    uses: ownable
+
+    # does not export ownable.transfer_ownership!
+
+    pending_owner: address  # the pending owner in the 2-step transfer process
+
+    @deploy
+    def __init__():
+        self.pending_owner = empty(address)
+
+    @external
+    def begin_transfer(new_owner: address):
+        ownable._check_owner()
+
+        self.pending_owner = new_owner
+
+    @external
+    def accept_transfer():
+        assert msg.sender == self.pending_owner
+
+        self.owner = new_owner
+        self.pending_owner = empty(address)
+
+Here, the ``ownable_2step`` module does not want to seal off access to calling the ``ownable`` module's ``__init__()`` function. So, it utilizes the ``uses: ownable`` statement to get access to the ``ownable`` module's state, without the requirement to initialize it. Note that this is a valid module, but it is not a valid contract (that is, it cannot produce bytecode) because it does not initialize the ``ownable`` module. To make a valid contract, the user of the ``ownable_2step`` module would be responsible for initializing the ``ownable`` module themselves (as in the next section: :ref:`initializing dependencies <init-dependencies>`).
+
```

Comment:
Add a note that a stateless modules must not be declared via `uses`.

Reply:
also should be an error

Reply:
yup, sgtm

Reply:
https://github.com/vyperlang/vyper/issues/4050